odl/space/pspace.py

```diff
@@ -711,31 +711,60 @@ def __setitem__(self, indices, values):
                 for p, v in zip(indexed_parts, values):
                     p[:] = v
 
-    def __array__(self):
-        """An array representation of ``self``.
+    def asarray(self, out=None):
+        """Extract the data of this vector as a numpy array.
 
         Only available if `is_power_space` is True.
 
         The ordering is such that it commutes with indexing::
 
             np.asarray(self[ind]) == np.asarray(self)[ind]
 
+        Parameters
+        ----------
+        out : `numpy.ndarray`, optional
+            Array in which the result should be written in-place.
+            Has to be contiguous and of the correct dtype and
+            shape.
+
         Examples
         --------
         >>> spc = odl.ProductSpace(odl.rn(2), 3)
         >>> x = spc.one()
-        >>> np.asarray(x)
+        >>> x.asarray()
         array([[ 1.,  1.],
                [ 1.,  1.],
                [ 1.,  1.]])
         """
         if not self.space.is_power_space:
             return NotImplemented
```

Comment:
I think we should stop returning `NotImplemented` here because that will result in just the same cryptic error message when people call `np.asarray(x)` and `x` is not a power space element. In this sense the PR does not really address (all of) the original issue it claims to close.

We should raise a proper exception with an explanation.

Reply:
Nice catch, I'll get this done.